pkg/registries/registries.go

```diff
@@ -71,55 +76,127 @@ func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy
 	}
 	sort.Strings(sources)
 	// Convert the sets of mirrors
-	res := []apioperatorsv1alpha1.RepositoryDigestMirrors{}
+	res := []apicfgv1.ImageTagMirrors{}
 	for _, source := range sources {
-		ds := disjointSets[source]
-		topoGraph := newTopoGraph()
-		for _, set := range *ds {
-			for i := 0; i+1 < len(set.Mirrors); i++ {
-				topoGraph.AddEdge(set.Mirrors[i], set.Mirrors[i+1])
+		sortedRepos, err := topoSortRepos(source, disjointSets[source])
+		if err != nil {
+			return nil, err
+		}
+		var mirrors []apicfgv1.ImageMirror
+		for _, repo := range sortedRepos {
+			mirrors = append(mirrors, apicfgv1.ImageMirror(repo))
+		}
+		imageTagMirror := apicfgv1.ImageTagMirrors{
+			Source:  source,
+			Mirrors: mirrors,
+		}
+		if mirrorBlockSource[source] {
+			imageTagMirror.MirrorSourcePolicy = apicfgv1.NeverContactSource
+		}
+		res = append(res, imageTagMirror)
+	}
+	return res, nil
+}
+
+// mergedDigestMirrorSets processes idmsRules and returns a set of ImageDigestMirrors, one for each Source value,
+// ordered consistently with the preference order of the individual entries (if possible)
+// E.g. given mirror sets (B, C) and (A, B), it will combine them into a single (A, B, C) set.
+func mergedDigestMirrorSets(idmsRules []*apicfgv1.ImageDigestMirrorSet) ([]apicfgv1.ImageDigestMirrors, error) {
```

Comment:
At a first glance it seems to me that it should be possible to share more of the logic, but I haven’t read the code that carefully. (I do appreciate there’s a `topoSortRepos` now.)

_Maybe_ an object that maintains `disjointSets`/`mirrorBlockSource`, with methods very vaguely like `addMirrorSet(source, neverContactSource, mirrors)` and `mergedMirrorSets() []struct{source, neverContactSource, mirrors}`? That way the two functions here would deal _only_ with the data conversion with no logic at all.

Or _maybe_, instead of `mergedMirrorSets`, a `sourceList() []string` and `updateRegistry(source, *Registry)`, to share more of the code in `EditRegistriesConfig` that just consumes the data from `mergedMirrorSets`.